### PR TITLE
[SDK:DDK][NTOS:PNP] Implement PnP arbiters initialization

### PR DESCRIPTION
**Pull request by vgalnt**
Implementation of the ArbInitializeArbiterInstance() function for the "arbiter" library.

### CODE REVIEW

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 1]
```diff
@@ -34,10 +230,135 @@ ArbInitializeArbiterInstance(
...
+        ExFreePoolWithTag(Arbiter->Allocation, 'RbrA');
+    }
+    if (Arbiter->PossibleAllocation)
+    {
+        ExFreePoolWithTag(Arbiter->PossibleAllocation, 'RbrA');
+    }
+    if (Arbiter->AllocationStack)
+    {
+        ExFreePoolWithTag(Arbiter->AllocationStack, 'AbrA');
+    }
```
Comment (Extravert-ir) [inline sdk/lib/drivers/arbiter/arbiter.c:360]: No need to check pointers here. If we reached this, all of them are initialized

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 2]
```diff
@@ -34,10 +230,135 @@ ArbInitializeArbiterInstance(
...
+    ASSERT(Arbiter->UnpackRequirement != NULL);
+    ASSERT(Arbiter->PackResource != NULL);
+    ASSERT(Arbiter->UnpackResource != NULL);
+
+    ASSERT(Arbiter->MutexEvent == NULL &&
+           Arbiter->Allocation == NULL &&
+           Arbiter->PossibleAllocation == NULL &&
+           Arbiter->AllocationStack == NULL);
+
+    Arbiter->Signature = 'sbrA';
```
Comment (Extravert-ir) [inline sdk/lib/drivers/arbiter/arbiter.c:244]: Please, create defines for signatures and allocation tags

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 3]
```diff
@@ -34,10 +230,135 @@ ArbInitializeArbiterInstance(
...
+    }
+
+    RtlInitializeRangeList(Arbiter->Allocation);
+    RtlInitializeRangeList(Arbiter->PossibleAllocation);
+
+    Arbiter->Name = ArbiterName;
+    Arbiter->ResourceType = ResourceType;
+    Arbiter->TransactionInProgress = FALSE;
+
+    if (!Arbiter->TestAllocation)
```
Comment (Extravert-ir) [inline sdk/lib/drivers/arbiter/arbiter.c:292]: Why checking these? Isn't this function supposed to be called once for each arbiter instance?

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 4]
```diff
@@ -22,6 +22,202 @@
...
+NTSTATUS
+NTAPI
+ArbRollbackAllocation(
+    _In_ PARBITER_INSTANCE Arbiter)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+/*  Not correct yet, FIXME! */
```
Comment (Extravert-ir) [inline sdk/lib/drivers/arbiter/arbiter.c:63]: What's not correct?

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 5]
```diff
@@ -22,6 +22,202 @@
 
 /* FUNCTIONS ******************************************************************/
 
+NTSTATUS
+NTAPI
+ArbTestAllocation(
```
Comment (Extravert-ir) [inline sdk/lib/drivers/arbiter/arbiter.c:27]: Would be nice to have some comment about the purpose of each function

File: ntoskrnl/io/pnpmgr/arbs.c [hunk 6]
```diff
@@ -92,17 +343,67 @@ IopMemInitialize(VOID)
 
     DPRINT("IopRootMemArbiter %p\n", &IopRootMemArbiter);
 
+    IopRootMemArbiter.UnpackRequirement = IopGenericUnpackRequirement;
+    IopRootMemArbiter.PackResource = IopGenericPackResource;
+    IopRootMemArbiter.UnpackResource = IopGenericUnpackResource;
+    IopRootMemArbiter.ScoreRequirement = IopGenericScoreRequirement;
+
+    IopRootMemArbiter.FindSuitableRange = IopMemFindSuitableRange;
```
Comment (Extravert-ir) [inline ntoskrnl/io/pnpmgr/arbs.c:351]: Why arbiters have different set of functions assigned?

**Review by Extravert-ir — CHANGES_REQUESTED** [review]
Looks fine in general

**Comment by Extravert-ir** [comment]
I would also like to rename the file `io/pnpmgr/arbs.c` -> `io/pnpmgr/arbiters.c`

File: sdk/lib/drivers/arbiter/arbiter.c (hunk 1 above)
Comment (vgalnt) [reply sdk/lib/drivers/arbiter/arbiter.c:360]: Ok

File: sdk/lib/drivers/arbiter/arbiter.c (hunk 2 above)
Comment (vgalnt) [reply sdk/lib/drivers/arbiter/arbiter.c:244]: Ok

File: sdk/lib/drivers/arbiter/arbiter.c (hunk 3 above)
Comment (vgalnt) [reply sdk/lib/drivers/arbiter/arbiter.c:292]: Assigned globally here if not assigned in an instance of an arbiter.

File: sdk/lib/drivers/arbiter/arbiter.c (hunk 4 above)
Comment (vgalnt) [reply sdk/lib/drivers/arbiter/arbiter.c:63]: https://github.com/reactos/reactos/pull/2471#discussion_r399233131

File: ntoskrnl/io/pnpmgr/arbs.c (hunk 6 above)
Comment (vgalnt) [reply ntoskrnl/io/pnpmgr/arbs.c:351]: For 'MEM' and 'PORT' arbiters, some functions are common (they have the same functionality).

File: sdk/lib/drivers/arbiter/arbiter.c [hunk 7]
```diff
@@ -34,10 +230,118 @@ ArbInitializeArbiterInstance(
 {
     NTSTATUS Status;
```
Comment (disean) [inline sdk/lib/drivers/arbiter/arbiter.c:232]: Is it possible to put the arbiter functions into PAGE section?

**Review by Extravert-ir — APPROVED** [review]

**Pull request merged**